File: Mootata/34회차/2210.숫자판 점프.py
dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]

board = [list(input().split()) for _ in range(5)]
answer = set()  # 중복 제거를 위한 set


def dfs(x, y, num):
    if len(num) == 6:  # 한칸씩 이동하면서 만든 숫자의 길이가 6이 되면 answer에 추가하고 멈춤
        answer.add(num)
        return

    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]

        if nx < 0 or ny < 0 or nx >= 5 or ny >= 5:  # 범위 밖 벗어나면 패스
            continue

        dfs(nx, ny, num + board[nx][ny])


for i in range(5):  # 임의의 위치에서 시작해야하므로,
    for j in range(5):  # 모든 위치에서 시작
        dfs(i, j, board[i][j])

print(len(answer))

File: apua34/2회차/2210.숫자판점프.js
/**
 * @file 2210.숫자판점프.js
 * @author @apua34
 * @brief 
 * @link https://www.acmicpc.net/problem/2210
 * @version 0.1
 * @date 2023-07-23
 * @strategy DFS
 *
 * @copyright Copyright (c) 2023 @apua34
 *
 */

const fs = require("fs");
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
const input = fs.readFileSync(filePath)
    .toString()
    .trim()
    .split("\n")

const inputArr = input.map((item) => item.split(' ').map(Number));

const answer = [];


function solution(x, y, str){
    const dir  = [
        [1, 0], 
        [-1, 0],
        [0, -1],
        [0, 1],
    ]

    if(str.length == 6){ // 문자열이 6자리가 되면 추가
        answer.push(str) // 1개씩
        return;
    }

    dir.forEach((target)=>{
        const [a, b] = target;

        // 상하좌우 이동이 가능한지,
        const top = x + a >= 0 
        const bottom = x + a < 5
        const left = y + b >= 0
        const right = y + b < 5

        if(top && bottom && left && right){
            const topBottm = x + a;
            const leftRight = y + b;
            const move = str + inputArr[x + a][y + b];
            solution(topBottm, leftRight, move);
        }
    })
}

for(let i = 0; i < 5; i++){
    for(let j = 0; j < 5; j++){
      solution(i, j , '') // START
    }
  }


console.log([...new Set(answer)].length)

